javascript/ql/lib/semmle/javascript/NPM.qll

```diff
@@ -12,8 +12,29 @@ class PackageJson extends JsonObject {
     this.isTopLevel()
   }
 
-  /** Gets the name of this package. */
-  string getPackageName() { result = this.getPropStringValue("name") }
+  /**
+   * Gets the name of this package.
+   * If the package is located under the package `pkg1` and its relative path is `foo/bar`, it can be `pkg1/foo/bar`.
+   */
+  string getPackageName() {
+    result = this.getPropStringValue("name")
+    or
+    exists(PackageJson parentPkg, Container currentDir, Container parentDir, string parentPkgName |
+      currentDir = this.getJsonFile().getParentContainer() and
+      parentDir = parentPkg.getJsonFile().getParentContainer() and
+      parentPkgName = parentPkg.getPropStringValue("name") and
+      (
+        parentDir.getParentContainer().getBaseName() = "node_modules"
+        or
+        // Scoped package is located in node_modules/@scope/pkgname
+        parentDir.getParentContainer().getParentContainer().getBaseName() = "node_modules" and
+        exists(parentPkgName.indexOf("/"))
+      ) and
```

Comment:
```suggestion
      parentPkgName = parentPkg.getPropStringValue("name") and
```

On further thought (sorry for the many rounds of feedback).  
Do we need to restrict to just packages contained within a `node_modules` folder?  

Monorepos that contain many packages could experience this pattern without a `node_module` folder being part of it?

Reply:
I think developers need to define `exports` property in `package.json` to access other packages, if they aren't located in `node_modules`.
I think removing these codes might cause some false positives, but they might be rare case, so we might be able to accept them...?

I looked at some packages and in the case of `@github/webauthn-json`, they defined `exports` property so that `@github/webauthn-json/extended` will work.

https://github.com/github/webauthn-json/blob/e932b3585fa70b0bd5b5a4012ba7dbad7b0a0d0f/package.json#L18
https://nodejs.org/api/packages.html#exports-sugar

If we remove the exports property, `@github/webauthn-json/extended` will not work.

```shell
$ node -e 'console.log(require("@github/webauthn-json"))'
{
  create: [Getter],
  get: [Getter],
  schema: [Getter],
  supported: [Getter]
}
$ gsed -i 's/exports/foobar/' package.json
$ node -e 'console.log(require("@github/webauthn-json"))'
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '@github/webauthn-json'
```


In https://github.com/keystonejs/keystone, we can call subpath modules like `@keystone-6/core/admin-ui/router` after `npm install`.
In this case, npm workspaces put the symlinks in the `node_modules` directory, so this PR can detect it.

https://github.com/keystonejs/keystone/blob/5ac8ef453/package.json#L109
https://docs.npmjs.com/cli/v7/using-npm/workspaces

Reply:
> I think developers need to define exports property in package.json to access other packages, if they aren't located in node_modules. 

We already have some support for `"exports"` elsewhere.  

---- 

My thinking was that we most of the time ignore `node_modules` folder when we parse a project, so your contribution is very rarely relevant.  

> but they might be rare case, so we might be able to accept them...? 

I'll run an evaluation if you try to remove the `node_modules` restriction.  
If the results are noisy (or if the performance goes bad), then we can revert the change again.

Reply:
The failing tests (see the CI below) revealed something interesting.  

We get some package names that include `node_modules`.  
E.g. the `Dependencies.ql` test finds a new package-name: `test-package/node_modules/bundled-package`.   
I don't think that was what you meant with this contribution, so maybe filter those out?  

I did some evaluations (internal links: [one](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-12975-f6a8cd__nightly-old__code-scanning/reports), [two](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/pr-12975-f6a8cd__default__code-scanning/reports)), and that looks fine.  
Nothing happened in those evaluations, everything is the same.

Reply:
Oh I see.
I'll filter them out.

Reply:
It looks fine now?

Reply:
It looks OK.  
I'll run an evaluation on it later, but our system for doing so is having some issues right now. 
And I'm on PTO thursday+friday, so it might be until early next week before I can check it out.  
Don't worry, I haven't forgotten about your PR.

Reply:
Thanks!
Enjoy your holiday! 😀